### DISCUSSION TITLE
Track Idle User Session - PTC Thingworx uses Rhino javascript engine

### DISCUSSION THREAD

**Discussion by gvenkata07** [mozilla/rhino#1030] [Q&A] [answered]
Can someone please help me on how to implement a custom JS code to track idle user session and alert users with a dialog popup, few minutes prior to session timeout. I'm working on this task for customizing our web app built using a PTC IoT product called 'Thingworx' and it uses Rhino javascript engine to interpret the custom JS code we write in it's composer. 

Following functions in regular JS are of no use and this PTC thingworx composer is unable to interpret them (Basically Rhino engine is unable to interpret them)
"documen.onclick/document.onmousemove/window.setInterval/console.log"

Any help is greatly appreciated, thanks.

**Comment by gvenkata07** [comment]
@ibukanov, @norrisboyd, @rbri  can one of you please look into this ?

**Comment by p-bakker** [comment, marked as answer]
> using a PTC IoT product called 'Thingworx' and it uses Rhino javascript engine to interpret the custom JS code we write in it's composer

This suggests to me that you should be directing these questions at Thingworx, as they apparently embed Rhino for you and might have specific requirements how to use it 

> "documen.onclick/document.onmousemove/window.setInterval/console.log"

This suggests to me you're looking for a browser-like environment, because all the things you mention there are browser-specific and don't exist outside the scope of a browser (unless whomwever embeds Rhino (like Thingworx) explicitly adds them
